src/ui/public/courier/fetch/call_client.js

```diff
@@ -79,65 +91,90 @@ export function CallClientProvider(Private, Promise, es) {
       });
     });
 
+    const searchStrategiesWithRequests = assignSearchRequestsToSearchStrategies(requestsToFetch);
+
     // We're going to create a new async context here, so that the logic within it can execute
     // asynchronously after we've returned a reference to defer.promise.
     Promise.resolve().then(async () => {
-      // Flatten the searchSource within each searchRequest to get the fetch params,
-      // e.g. body, filters, index pattern, query.
-      const allFetchParams = await getAllFetchParams(requestsToFetch);
+      // Execute each request using its search strategy.
+      for (let i = 0; i < searchStrategiesWithRequests.length; i++) {
+        const searchStrategyWithSearchRequests = searchStrategiesWithRequests[i];
+        const { searchStrategy, searchRequests } = searchStrategyWithSearchRequests;
+        const {
+          searching,
+          abort,
+          failedSearchRequests,
+        } = await searchStrategy.search({ searchRequests, es, Promise, serializeFetchParams });
+
+        // Collect searchRequests which have successfully been sent.
+        searchRequests.forEach(searchRequest => {
+          if (failedSearchRequests.includes(searchRequest)) {
+            return;
+          }
 
-      // Serialize the fetch params into a format suitable for the body of an ES query.
-      const serializedFetchParams = await serializeAllFetchParams(allFetchParams, requestsToFetch);
+          activeSearchRequests.push(searchRequest);
+        });
 
-      // The index of the request inside requestsToFetch determines which response is mapped to it.
-      // If a request won't generate a response, since it already failed, we need to remove the
-      // request from the requestsToFetch array so the indexes will continue to match up to the
-      // responses correctly.
-      requestsToFetch = requestsToFetch.filter(request => request !== undefined);
+        abortableSearches.push({
+          searching,
+          abort,
+        });
+      }
 
       try {
         // The request was aborted while we were doing the above logic.
-        if (isRequestAborted) {
-          throw ABORTED;
+        if (areAllSearchRequestsAborted) {
+          return;
         }
 
-        esPromise = es.msearch({ body: serializedFetchParams });
-        const clientResponse = await esPromise;
-        await respond(clientResponse.responses);
-      } catch(error) {
-        if (error === ABORTED) {
-          return await respond();
-        }
+        const segregatedResponses = await Promise.all(abortableSearches.map(({ searching }) => searching));
+
+        // Assigning searchRequests to strategies means that the responses come back in a different
+        // order than the original searchRequests. So we'll put them back in order so that we can
+        // use the order to associate each response with the original request.
+        const responsesInOriginalRequestOrder = new Array(searchRequestsAndStatuses.length);
```

Comment:
Nothing "wrong" with this here, but I think it would be logical to use the `searchRequests` array instead of the `searchRequestsAndStatuses` array here, below, and in `respondToSearchRequests()`, don't you think? Especially since `respondToSearchRequests()` is reading the status from `searchRequestsAndStatuses` by index anyway.

Reply:
At least for me, using `searchRequestsAndStatuses` makes the code a bit easier to follow by reducing the number of variables I have to keep in mind. In terms of data structures being consumed, `searchRequests` stops being important immediately, and `searchRequestsAndStatuses`, `requestsToFetch`, and `searchStrategiesWithRequests` become the important ones. It would be a bit confusing to me if I were to read this code and see `searchRequests` was reintroduced -- I'd wonder what the intention behind that was.